src/daft-core/src/array/ops/clip.rs

```diff
@@ -0,0 +1,106 @@
+use arrow2::array::PrimitiveArray;
+use common_error::{DaftError, DaftResult};
+use num_traits::{clamp, clamp_max, clamp_min};
+
+use crate::{array::DataArray, datatypes::DaftNumericType, prelude::AsArrow};
+
+/// Helper function to perform clamping on a value with optional bounds using `Option<T>` directly
+fn clamp_helper<T: PartialOrd + Copy>(
+    value: Option<&T>,
+    left_bound: Option<&T>,
+    right_bound: Option<&T>,
+) -> Option<T> {
+    match (value, left_bound, right_bound) {
+        (None, _, _) => None,
+        (Some(v), Some(l), Some(r)) => {
+            assert!(l <= r, "Left bound is greater than right bound");
+            Some(clamp(*v, *l, *r))
+        }
+        (Some(v), Some(l), None) => Some(clamp_min(*v, *l)),
+        (Some(v), None, Some(r)) => Some(clamp_max(*v, *r)),
+        (Some(v), None, None) => Some(*v),
+    }
+}
+
+macro_rules! create_data_array {
```

Comment:
I don't see much benefit in this macro, since the amount of lines covered is pretty minimal.